### PR TITLE
Feat: filter definitions by which addon installed them

### LINKED ISSUE #4140: [Feature] add label to definition to identify which addon install them
**Is your feature request related to a problem? Please describe.**

More and more definitions are installed with addon enabled, we can hardly check every definition by list all of them.

I hope we can have some convenience way to list , such as:

```
vela def list --from fluxcd
```

Then it will list all definitions from fluxcd addon.

**Describe the solution you'd like**

1. add a label to all definitions enabled by addon, e.g. `addon.oam.dev: fluxcd`
2. select label when users run `vela def list --from fluxcd`

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by charlie0129**
Signed-off-by: Charlie Chiang <charlie_c_0129@outlook.com>

### Description of your changes

- In `vela def list`, now we can filter definitions by which addon installed them.
- APIServer is also updated with a `ownerAddon` query parameter in `/api/v1/definitions`. Swagger doc updates included.

I used ownReference to find out which addon installed them, which requires less code changes.

e.g. `vela def list --from fluxcd` only shows definitions that are installed by `fluxcd`

<details>

```console
$ vdl --from fluxcd
NAME                    	TYPE               	NAMESPACE  	DESCRIPTION                                                 
kustomize-json-patch    	TraitDefinition    	vela-system	A list of JSON6902 patch to selected target                 
kustomize-patch         	TraitDefinition    	vela-system	A list of StrategicMerge or JSON6902 patch to selected      
                        	                   	           	target                                                      
kustomize-strategy-merge	TraitDefinition    	vela-system	A list of strategic merge to kustomize config               
config-helm-repository  	ComponentDefinition	vela-system	Config information to authenticate helm chart repository    
helm                    	ComponentDefinition	vela-system	helm release is a group of K8s resources from either git    
                        	                   	           	repository or helm repo                                     
kustomize               	ComponentDefinition	vela-system	kustomize can fetching, building, updating and applying     
                        	                   	           	Kustomize manifests from git repo.
```
</details>

Screenshot from 2022-06-12 01-30-13

Fixes #4140

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

relevant unit tests have been added

### Special notes for your reviewer

Core changes are in:
- `pkg/definition/definition.go`
- `pkg/utils/filters/filter.go`
- `references/cli/def.go`
- `pkg/apiserver/domain/service/definition.go`

The reason why there is so many changes is that I refactored two helper functions, moving them from a generic package to a helper package. This is required to prevent import-cycles and will make the code more organized.

### CODE REVIEW

File: pkg/definition/definition.go [hunk 1]
```diff
@@ -375,7 +377,7 @@ func ValidDefinitionTypes() []string {
 }
 
 // SearchDefinition search the Definition in k8s by traversing all possible results across types or namespaces
-func SearchDefinition(definitionName string, c client.Client, definitionType string, namespace string) ([]unstructured.Unstructured, error) {
+func SearchDefinition(definitionName string, c client.Client, definitionType string, namespace string, addonFilter string) ([]unstructured.Unstructured, error) {
```
Comment (wonderflow) [inline pkg/definition/definition.go:380]: the addonFilter can be a function interface, then it's portable to add other filter for extensibility:

```suggestion
func SearchDefinition(definitionName string, c client.Client, definitionType, namespace string, addonFilter func(unstructured.Unstructured)bool) ([]unstructured.Unstructured, error) {
```

**Review by wonderflow — COMMENTED** [review]
I'm really glad that you didn't even change the definintion created by using the ownerReference to search! Your solution is better than my proposal in the issue!

Really great work! 👍

File: pkg/definition/definition.go (hunk 1 above)
Comment (charlie0129) [reply pkg/definition/definition.go]: Great, I refactored the code to accept function interface as generic filters, and encapsulated the original implementation into one of the filters.

Comment (charlie0129) [reply pkg/definition/definition.go]: Well, in the original function parameters `SearchDefinition(definitionName string, c client.Client, definitionType, namespace string)`, like `definitionType`, `namespace`, and `definitionName` can all be replaced with filters, in unified manner. I wonder if this refactor is necessary.

**Comment by barnettZQG** [comment]
This is a great feature, I have more advice:

1. Show the addon name in the def list.
2. Add the same feature in the API of listing definitions.

**Comment by charlie0129** [comment]
> This is a great feature, I have more advice:
> 
> 1. Show the addon name in the def list.
> 2. Add the same feature in the API of listing definitions.

Sure, I will implement them later!

File: pkg/definition/definition.go (hunk 1 above)
Comment (wonderflow) [reply pkg/definition/definition.go]: @charlie0129 You're really smart! That's exactly what I mean. Please make sure they all covered by tests.

File: pkg/apiserver/domain/service/definition.go [hunk 2]
```diff
@@ -63,12 +65,13 @@ type definitionServiceImpl struct {
...
 	Type             string `json:"type"`
 	AppliedWorkloads string `json:"appliedWorkloads"`
+	OwnerAddon       string `json:"sourceAddon"`
 	QueryAll         bool   `json:"queryAll"`
 }
 
 // String return cache key string
 func (d DefinitionQueryOption) String() string {
-	return fmt.Sprintf("type:%s/appliedWorkloads:%s/queryAll:%v", d.Type, d.AppliedWorkloads, d.QueryAll)
+	return fmt.Sprintf("type:%s/appliedWorkloads:%s/ownerAddon:%s/queryAll:%v", d.Type, d.AppliedWorkloads, d.OwnerAddon, d.QueryAll)
```
Comment (wonderflow) [inline pkg/apiserver/domain/service/definition.go:74]: will it break the compatibility before?

Comment (wonderflow) [reply pkg/apiserver/domain/service/definition.go:74]: \cc @barnettZQG @wangyikewxgm

File: pkg/utils/filters/filter.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+*/
+
+package filters
+
+import (
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
+
```
Comment (wonderflow) [inline pkg/utils/filters/filter.go:24]: remove the empty line, please format the headers.

File: pkg/utils/filters/filter_test.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
+
+	"gotest.tools/assert"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	addonutil "github.com/oam-dev/kubevela/pkg/utils/addon"
```
Comment (wonderflow) [inline pkg/utils/filters/filter_test.go:31]: format

File: references/cli/def.go [hunk 5]
```diff
@@ -30,6 +30,12 @@ import (
 	"strings"
 	"time"
 
+	addonutil "github.com/oam-dev/kubevela/pkg/utils/addon"
+
+	"github.com/oam-dev/kubevela/pkg/utils/filters"
+
+	"github.com/oam-dev/kubevela/pkg/cue/packages"
```
Comment (wonderflow) [inline references/cli/def.go:37]: format

File: pkg/utils/filters/filter.go (hunk 3 above)
Comment (charlie0129) [reply pkg/utils/filters/filter.go]: well, I didn't notice that. just cannot fully rely goimports

File: pkg/apiserver/domain/service/definition.go (hunk 2 above)
Comment (charlie0129) [reply pkg/apiserver/domain/service/definition.go:74]: AFAIK, it is only used in caches (when I wrote this code 16 days ago). Now, the cache is removed #4147. So this function is not even used (from my initial inspections).

File: pkg/utils/filters/filter_test.go (hunk 4 above)
Comment (charlie0129) [reply pkg/utils/filters/filter_test.go:28]: It's strange that my changes didn't show up here. The other two have.

Just wait to see...

File: pkg/apiserver/domain/service/definition.go (hunk 2 above)
Comment (wonderflow) [reply pkg/apiserver/domain/service/definition.go:74]: Great

File: pkg/apiserver/domain/service/definition.go [hunk 6]
```diff
@@ -119,24 +122,17 @@ func (d *definitionServiceImpl) listDefinitions(ctx context.Context, list *unstr
...
-					filter = true
-					break
-				}
-			}
-			if !filter {
-				continue
-			}
-		}
+	for _, def := range filteredList.Items {
 		definition, err := convertDefinitionBase(def, kind)
```
Comment (barnettZQG) [inline pkg/apiserver/domain/service/definition.go:136]: Add the OwnerAddon field in the definition base struct.

**Review by Somefive — APPROVED** [review]

**Review by barnettZQG — APPROVED** [review]

**Pull request merged**